core/src/main/java/de/danoeh/antennapod/core/sync/SyncService.java

```diff
@@ -69,29 +81,38 @@
     private static final String WORK_ID_SYNC = "SyncServiceWorkId";
     private static final ReentrantLock lock = new ReentrantLock();
 
-    private ISyncService syncServiceImpl;
+    private final Map<String, ISyncService> syncServices = new HashMap<>();
 
     public SyncService(@NonNull Context context, @NonNull WorkerParameters params) {
         super(context, params);
+        syncServices.put(SYNC_PROVIDER_CHOICE_GPODDER_NET, new GpodnetService(AntennapodHttpClient.getHttpClient(),
+                GpodnetPreferences.getHosturl(), GpodnetPreferences.getDeviceID(),
+                GpodnetPreferences.getUsername(), GpodnetPreferences.getPassword()));
+
+        syncServices.put(
+                SYNC_PROVIDER_CHOICE_NEXTCLOUD,
+                new NextcloudSyncService(getApplicationContext(), ClientConfig.USER_AGENT)
+        );
     }
 
     @Override
     @NonNull
     public Result doWork() {
-        if (!GpodnetPreferences.loggedIn()) {
-            return Result.success();
-        }
-        syncServiceImpl = new GpodnetService(AntennapodHttpClient.getHttpClient(),
-                GpodnetPreferences.getHosturl(), GpodnetPreferences.getDeviceID(),
-                GpodnetPreferences.getUsername(), GpodnetPreferences.getPassword());
         SharedPreferences.Editor prefs = getApplicationContext().getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
                 .edit();
+        String selectedService = getSelectedSyncProviderKey(getApplicationContext());
+        ISyncService activeSyncProvider = getActiveSyncProvider(selectedService, this.syncServices);
+
+        return getResultForService(prefs, activeSyncProvider);
+    }
+
+    private Result getResultForService(SharedPreferences.Editor prefs, ISyncService syncServiceImpl) {
```

Comment:
When pressing sync for the first time, I see "sync failed" with the current date during sync. Could this be related to the following stack trace?

```
2021-08-03 21:07:00.534 31039-31064/de.danoeh.antennapod.debug E/WM-WorkerWrapper: Work [ id=018c55d9-88d6-4489-b77b-25a44e741720, tags={ de.danoeh.antennapod.core.sync.SyncService } ] failed because it threw an exception/error
    java.util.concurrent.ExecutionException: java.lang.NullPointerException: Attempt to invoke interface method 'void de.danoeh.antennapod.net.sync.model.ISyncService.login()' on a null object reference
        at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:516)
        at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
        at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:298)
        at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:923)
     Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'void de.danoeh.antennapod.net.sync.model.ISyncService.login()' on a null object reference
        at de.danoeh.antennapod.core.sync.SyncService.getResultForService(SyncService.java:112)
        at de.danoeh.antennapod.core.sync.SyncService.doWork(SyncService.java:106)
        at androidx.work.Worker$1.run(Worker.java:85)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:923) 
2021-08-03 21:07:00.541 31039-31064/de.danoeh.antennapod.debug I/WM-WorkerWrapper: Worker result FAILURE for Work [ id=018c55d9-88d6-4489-b77b-25a44e741720, tags={ de.danoeh.antennapod.core.sync.SyncService } ]
```

Reply:
i added the early exit again